Semgrep/Azure AD SAML SSO guide (#786)
returntocorp/semgrep-docs: docs/semgrep-app/sso.md

If you encounter issues during the setup process, reach out to [support@r2c.dev](mailto:support@r2c.dev) for assistance.

### Setting up SAML SSO with Azure Active Directory

Set up Semgrep as an Enterprise App with SAML SSO using Azure's Active Directory service by following these steps:

> 1. Enable **Enterprise App** in **Azure Active Directory** with SAML SSO selected. Follow the instructions in [Enable single sign-on for an enterprise application](https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/add-application-portal-setup-sso).

Reviewer: Which section or sections of the linked document should we read to find this information?

Reviewer 2: outdated - resolving as "addressed in subsequent commits by me"

> 2. Go to Semgrep App [Settings](https://semgrep.dev/orgs/-/settings/sso) page. TODO Under **Basic Configuration**, you'll need to enter the `Entity ID` and the `Reply URL`. These are provided in the Semgrep App on the left-side **Settings** tab, under the **SSO** tab on the top, as `Audience URL` and `SSO URL`, respectively. Make sure you switch the provider type to **SAML2** in the SSO form in the Semgrep App.

Reviewer: I don't see **Basic Configuration** in the Semgrep App. Do we mean **Basic SAML Configuration** on the side of Azure? We need to specify at the beginning of each step where we are. These steps are too long, there should be one or two actions at maximum for a user to do in each step.

    !["Entity ID" and "Reply URL" fields in Semgrep App](/img/azure-sso-fields-2.png ""Entity ID" and "Reploy URL" fields from within the App")

> 3. TODO You need to add two attribute statements called `name` and `email`, making sure to set the `email` attribute to `user.mail`. For help regarding **Attributes and Claims**, refer to [Customize claims issued in the SAML token for enterprise applications](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-saml-claims-customization) documentation.

Reviewer: Great that we are linking to specific docs. Which section or sections of the linked document should we read to find this information?

Reviewer 2: resolving as "addressed in subsequent commits by me"

> 4. In the Azure Single sign-on page, find the `Login URL`, `Azure AD Identifier`, and `Certificate (Base64)`. Pate them to your [Settings](https://semgrep.dev/orgs/-/settings/sso) page of SAML2 SSO as `IdP SSO URL` (`Login URL`), `IdP Issuer ID` (`Azure AD Identifier`), and `X.509 Certificate` (`Certificate (Base64)`). 

Reviewer: ```suggestion
4. In the Azure Single sign-on page, find the `Login URL`, `Azure AD Identifier`, and `Certificate (Base64)`. Paste them to your [Settings](https://semgrep.dev/orgs/-/settings/sso) page of SAML2 SSO as `IdP SSO URL` (`Login URL`), `IdP Issuer ID` (`Azure AD Identifier`), and `X.509 Certificate` (`Certificate (Base64)`). 
```

    ![Setting additional SSO form fields](/img/azure-sso-app-fields.png "Setting additional SSO form fields")

> 5. TODO At the top of the form, fill in your email domain and a display name for the SSO integration.

Reviewer: This seems to be prefilled already (I checked in two organizations).

Reviewer 2: FYI, on my end it was not prefilled @adamkvitek . Probably depends on the version of Semgrep App when the org was added.
<img width="581" alt="2022-12-09 20_02" src="https://user-images.githubusercontent.com/20766840/206698492-d714d5c7-ee24-443f-a6ae-41b2a331117b.png">

    ![Display name and email domain in the SSO form](/img/azure-sso-app-fields-3.png "Display name and email domain in the SSO form")
6. Once all the fields are filled out, press the **Create/Update Auth Provider** button at the bottom of the form. See a green confirmation message in the upper-left of the app to verify that we've recieved your SSO info.

>     ![Submitting your SSO info in the App](/img/azure-sso-submit.png "Submitting your SSO info in the App")

Reviewer: We do not have to include this screenshot...
```suggestion
```

> To add users to the application in so they can log in with their domain emails: refer to the [Assign users and groups to an application](https://learn.microsoft.com/en-us/azure/active-directory/manage-apps/assign-user-or-group-access-portal).

Reviewer: So this change is only done on the side of Azure? Our users do not have to enable anything in Semgrep App?

3. Click **New application** > **Create your own application**. A menu appears.
![Screenshot of create your own Application page in Azure AD](/img/azure-create-application.png#bordered)
4. Give your new application a name, such as `Semgrep SAML`.
5. Click **Integrate any other application you don't find in the gallery (non-gallery)**.
6. Click **Create**. This takes you to your new Enterprise Application's page.

You have now created a custom Enterprise App for Semgrep to integrate into Azure AD. This enables you to set up SAML SSO.

#### Setting up SAML SSO for your new Enterprise App

1. From your new Enterprise App's page, Click **Single-sign on** > **SAML** tile. This takes you to the **Set up Single Sign-On with SAML** page.
![Screenshot of your Enterprise App's page](/img/azure-select-sso.png#bordered)
2. Under the **Basic SAML Configuration** form, click **Edit** to enter values for **Entity ID** and **Reply URL**. These values are retrieved from Semgrep App by performing the following steps:
    1. Go to Semgrep App [Settings](https://semgrep.dev/orgs/-/settings/sso) page, and then click **SAML2** under **SSO Method** to display the SAML2 form.

>     2. Copy the **Audience URL** value from Semgrep App to **Entity ID**.

Reviewer: @s-santillan feel free to simplify my suggestion, I tried to do my best... but it is a complex UI.

>     3. Copy the **SSO URL** value from Semgrep App to **Reply URL**.

Reviewer: ```suggestion
    3. Copy the **SSO URL** value from Semgrep App, then in **Basic SAML Configuration**, under **Reply URL (Assertion Consumer Service URL)** click the **Add reply URL** and paste in the URL.
```

3. In the **Basic SAML Configuration** form, click **Save**.

> 4. Under the **User Attributes & Claims** form, click **Edit** > **Add new claim**.

Reviewer: ```suggestion
4. Under the **Attributes & Claims** form, click **Edit** > **Add new claim**.
```

    1. Enter `name` in the **Name** field.
    2. From the **Source attribute** drop-down box, select `user.givenname`.

>     ![Screenshot of create your own Application page in Azure AD](/img/azure_ad-add-claim.png#bordered)

Reviewer: ```suggestion
    ![Screenshot of create your own Application page in Azure AD](/img/azure_ad-add-claim.png#bordered)
    3. Click **Save**.
```

> 4. Under the **User Attributes & Claims** form, click **Edit** < **Add new claim**.

Reviewer: ```suggestion
4. Under the **Attributes & Claims** form, click **Edit** > **Add new claim**.
```

    1. Enter `email` in the **Name** field.

>     2. From the **Source attribute** drop-down box, select `user.email`.

Reviewer: ```suggestion
    2. From the **Source attribute** drop-down box, select `user.email`.
    3. Click **Save**.
```

6. Fill out the values required by Semgrep App's SAML2 form (**IdP SSO URL**, **IdP Issuer ID**, **X.509 Certificate**) by copying the values from Azure AD's **Set up Single Sign-On with SAML** page. Perform the following steps:

>     1. Copy the **Login URL** value from Azure AD to **IDP SSO URL**. 

Reviewer: ```suggestion
    1. Copy the **Login URL** value from Azure AD to **IDP SSO URL** in Semgrep App. 
```

>     2. Copy **Azure AD Identifier** value to **IdP Issuer ID**.

Reviewer: ```suggestion
    2. Copy **Azure AD Identifier** value to **IdP Issuer ID** in Semgrep App.
```

>     3. Click **Download** next to **Certificate (Base64)**.

Reviewer: ```suggestion
    3. In **Set up Single Sign-On with SAML**, click **Download** next to **Certificate (Base64)**.
```

>     4. Copy the contents of the downloaded certificate to **X.509 Certificate**. 

Reviewer: ```suggestion
    4. In Semgrep App, under **X.509 Certificate**, click **Browse** and then add the downloaded file. 
```
